### PR TITLE
Migrate to Terraform 0.12

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
## Description

This PR is based and supersedes https://github.com/kinvolk/lokomotive-kubernetes/pull/45. It migrates the different providers and updates the documentation accordingly.

I would need some help here to test the providers I already tested and also to test the ones I wasn't able to test yet. 

The commits are organized as follows:
- preparatory commits, includes also fixes for some already existing problems, like https://github.com/kinvolk/lokomotive-kubernetes/pull/116.
- run terraform 0.12upgrade, just run it without fixing anything.
- fixes the problems introduced by the above.
- update documentation / examples / ci.
- use some features of Terraform 0.12, like array notation and  specific data types.

### Status
- Aws,  Azure, Google Cloud and Packet providers have been migrated and tested.
- Baremetal and kvm-libvirt have been migrated but not tested yet.
- DigitalOcean hasn't been migrated.

### Testing

Given that it is a quite large PR, I'm adding this table to check that all plaforms are tested.
Please add your name (so we can blame you later on [not true :rofl: ]) when you test a platform.

| Platform | Tester 1 | Tester 2 | Tester 3? |
|------------ |-------------|------------|-------------|
| AWS | @mauriciovasquezbernal  | | | 
| Azure | @mauriciovasquezbernal  | | |
| Bare Metal | | | |
| Google-Cloud | @mauriciovasquezbernal  | | |
| KVM-libvirt | @ipochi | | |
| Packet | @mauriciovasquezbernal  | @johananl  | |

### TODO

- [ ] Update bootkube module source once https://github.com/kinvolk/terraform-render-bootkube/pull/33 is merged.
- [ ] Update CI.

### CODE REVIEW

**Comment by ipochi** [comment]
tested on kvm-libvirt, works fine

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (rata) [inline packet/flatcar-linux/kubernetes/bootkube.tf]: Was done by the 0.12 upgrade script or changed by you? (i mean the `git::https` prefix)

File: packet/flatcar-linux/kubernetes/controllers.tf
Comment (rata) [inline packet/flatcar-linux/kubernetes/controllers.tf:54]: is this new style generated by `terraform fmt` or the upgrade script? Or why are we changing the coding style?

I guess, for other changes, that either the upgrade script or terraform fmt does this now?

**Review by rata — COMMENTED** [review]
I think it would be easier to review if the manual changes and the changes done by the upgrade script are separated commits. If not, I usually don't know why a change was done. For example, I find myself when looking at the diff of one commit only, not knowing if we change something for the style, to use bool data type, or it was just part of the auto-generated code by the upgrade script (as the commit just list all of these)

The same applies for the "preparation" changes, like renaming `count` to `worker_count`. It is all in the same commit and if it wasn't because I upgraded another repo to terraform 0.12, I wouldn't know that this was warned by the terraform 0.11 upgrade check and required. I'd thought this is an unrelated improvement and not relevant for the PR, for example :-D

If the changes were split across several commits (like rename `count` to `worker_count` and explain why that is needed), then the commit would have to mention why a change was done and is, again IMHO, way more easy to review. Just go one commit at a time, and they kind of tell the story: rename x -> b because terraform warned about this, upgrade to terraform 0.12, change to use X, etc. But maybe this is personal taste.

But if, instead, the commit is doing several things is more difficult to catch mistakes (as you are looking to a big diff all at once) and usually don't explain why something was done. So, it is difficult to asses if that chunk is actually needed or not, or how it can be improved. IMHO

There doesn't seem to be so many changes mixed, though, so if it is difficult for you to split them now, I can review as it is too. But maybe consider it for the next time, if possible? Does it makes sense for you too or you did it like this on purpose? 

(btw, maybe is not difficult to split because you can go to the commit, just run the upgrade script, commit that, and then copy the resulting file with your change over that and commit, in the best way you consider? So, it is just a matter of explaining the commits and splitting, maybe?)

**Comment by mauriciovasquezbernal** [comment]
@rata I had saved a backup branch with more granular commits, I clean those up and pushed here. I hope this new version is easier to review.

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (mauriciovasquezbernal) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: `git::https` is needed because ref is a branch.

File: packet/flatcar-linux/kubernetes/controllers.tf
Comment (mauriciovasquezbernal) [reply packet/flatcar-linux/kubernetes/controllers.tf:54]: It's done by the `terraform 0.16upgrade`

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (rata) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Oh, really? That sound weird... Is this documented somewhere? Or you saw an error that it had problems to disambiguate or something?

File: packet/flatcar-linux/kubernetes/controllers.tf
Comment (rata) [reply packet/flatcar-linux/kubernetes/controllers.tf:54]: 0.12 I guess? Thanks!

File: packet/flatcar-linux/kubernetes/require.tf
Comment (rata) [inline packet/flatcar-linux/kubernetes/require.tf]: Why 2.2.1? Support for tf 0.12 was added in [2.1.0](https://github.com/terraform-providers/terraform-provider-packet/blob/master/CHANGELOG.md#210-april-30-2019)

If the idea is to use the oldest that supports terraform 0.12 why not using 2.1.0? In any case, the latest 2.7.3 seems bettern IMHO, as we are already doing all the work to test this, we will need to repeat this when upgrading the provider version to the latest later again, right?

Am I missing something?

File: azure/flatcar-linux/kubernetes/controllers.tf
Comment (rata) [inline azure/flatcar-linux/kubernetes/controllers.tf]: Ugh, this seems like something that might come to bite us later (node created with only one key, a user trying to ssh to debug a provisioning issue and they key not being the first, and therefore failing to ssh. And weird stuff like that)

Why provide only one key instead of all the keys? My gut feeling is that this probably was failing due to a formatting error (the `key_data` arg might be expecting this in another format that specified in the big terraform string?).

I'd like to understand better the implications of this change, and doesn't seem obvious at all for me. In fact, if I see that line of code I'd expect it to be a bug or a big fat comment explaining why it is not an issue or the trade-offs done.

Can you elaborate on why this is not an issue? Or maybe that is out of scope for this PR? In which case, I'd not add this commit.

Or does a deployment fail without this? And did it failed using the provider before this PR?

File: docs/flatcar-linux/kvm-libvirt.md
Comment (rata) [inline docs/flatcar-linux/kvm-libvirt.md:94]: nit: You can just use `tar xf`, it will discover if the file is gzip, bzip, etc. :)

File: docs/flatcar-linux/kvm-libvirt.md
Comment (rata) [inline docs/flatcar-linux/kvm-libvirt.md:81]: Was the ct provider wrongly specified before? Or is this needed to update to terraform 0.12?

It seems 0.3.2 added support for terraform 0.12 (https://github.com/poseidon/terraform-provider-ct/releases), so maybe this was fine just not supported terraform 0.12?

Maybe a commit title like:

```
docs/kvm-libvirt: Update instructions to use ct 0.4.0

Next commits will migrate this to terraform 0.12 and the ct provider should be updated to support terraform 0.12 too.
```

Does it make sense?

File: docs/flatcar-linux/bare-metal.md
Comment (rata) [inline docs/flatcar-linux/bare-metal.md]: Does this version work with terraform 0.11? If compatibility is not kept, maybe is better to do this atomically in the commit that upgrades to terraform 0.12. If not, the tree may be completely broken in this commit

File: aws/flatcar-linux/kubernetes/workers/variables.tf
Comment (rata) [inline aws/flatcar-linux/kubernetes/workers/variables.tf:25]: This is a breaking change for all the providers. It would be nice to list somewhere (PR description?) all the breaking changes needed to upgrade to 0.12 support (like this change :))? What do you think?

File: aws/flatcar-linux/kubernetes/bootkube.tf
Comment (rata) [inline aws/flatcar-linux/kubernetes/bootkube.tf]: The tree is probably broken before this commit. Does it make sense to you too to have the tree working at all commits and, therefore squash this with the upgrade to terraform 0.12? (maybe split by platforms if that makes it more easy to review?)

**Review by rata — COMMENTED** [review]
@mauriciovasquezbernal thanks a lot! It is indeed easier for me and I think it helped to catch some issues that I wouldn't catch without this. Thanks!

Did some small comments and a general comment: I think in this PR a commit does a change but leaves the repo in a state that doesn't work (i.e. a deployment or running the tests might fail). Does it make sense to try to not do that? IOW, have at every commit a working tree that you can, in theory, run the tests/deploy? That way it commit is atomic and doesn't break tests, etc. making bisects in the future easier. Does it make sense to you?

Like, in the current PR I think it is mostly about: keep the "prepatary" commits and split the terraform 0.12 upgrade per platform and commit each platform upgrade (runnning the terraform script) together with the fixes needed for that platform to be working again. And, then, at every commit it _should_ work.

IOW, I think if we just split per platform and squash all the fixes (with their explanations), we don't have commits that break things.

Does it make sense to you?

If it doesn't make sense to you or is too much work or something, don't hesitate to say it :). I don't want to be asking for changes that you don't see any value in doing them. I'm ok with this PR as it is split if you think it is not worth changing it, I think (will have to finish the review to be sure :)).

Will try to continue the review in the next working day or so :)

PS: I'm not saying you should test this at every commit, please don't! Just to not have a the repo obviously broken at any given commit. If we want to enforce this, the CI should do it in the future, of course :)

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (mauriciovasquezbernal) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: I saw an error and then have to change. Actually I'm not sure what is the reason for this, given that it's a temporal change I didn't go deep in the details.

File: packet/flatcar-linux/kubernetes/require.tf
Comment (mauriciovasquezbernal) [reply packet/flatcar-linux/kubernetes/require.tf]: You're right. 2.2.1 was a copy & paste from https://github.com/kinvolk/lokomotive-kubernetes/pull/45. I updated it to 2.7.3.

File: azure/flatcar-linux/kubernetes/controllers.tf
Comment (mauriciovasquezbernal) [reply azure/flatcar-linux/kubernetes/controllers.tf]: The Azure Terraform module requires a least one `ssh_keys` block when `disable_password_authentication` is `true`. After https://github.com/kinvolk/lokomotive-kubernetes/pull/98 I found out that it was not working with multiple keys, this commit provides a single key that is required. Please note that all the keys are copied by ignition anyway, the problem of only a single key working should never happen.

I don't think we need to put special attention to try to copy all the keys here too as they are already handled by Ignition, as stated in the small comment at https://github.com/kinvolk/lokomotive-kubernetes/blob/afc4e928449d44cd1b3d5f7b27f09ef238a41685/azure/flatcar-linux/kubernetes/controllers.tf#L73.
I'll add a further comment to clear it out.

File: docs/flatcar-linux/kvm-libvirt.md
Comment (mauriciovasquezbernal) [reply docs/flatcar-linux/kvm-libvirt.md:81]: Please notice that this commit just fixes a typo on the documentation. ct 0.4.0 was already used.

File: docs/flatcar-linux/bare-metal.md
Comment (mauriciovasquezbernal) [reply docs/flatcar-linux/bare-metal.md]: Still works with Terrraform 0.11: https://github.com/poseidon/terraform-provider-matchbox/blob/master/CHANGES.md#v030

**Comment by mauriciovasquezbernal** [comment]
@rata I think having a working tree on each commit and having small commits could be complicated. For instance, after running `terraform 0.12upgrade` it is needed to fix some issues by hand, if I do that in a single commit then it'll be not possible to understand what was changed by the tool and what was fixed manually.

I like the idea of having a working tree on all commits, but I'm don't know if it's worth to do so.

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (rata) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Oh, right, this is temporal. Never mind! :)

File: bare-metal/flatcar-linux/kubernetes/profiles.tf
Comment (rata) [inline bare-metal/flatcar-linux/kubernetes/profiles.tf]: Sorry, not sure what this means. Do you plan to address this in this PR? Or this will be addressed later?

File: aws/flatcar-linux/kubernetes/variables.tf
Comment (rata) [inline aws/flatcar-linux/kubernetes/variables.tf:113]: This very much depends on what type bootkube accepts, right? So, since the commit bootkube was changed till now it is also broken. Am I understanding correctly?

Just to make sure I follow, not asking for a change :)

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (johananl) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: The following works fine for me:

```hcl
module "tf-test" {
  source = "github.com/johananl/tf-test?ref=branch"
}
```

It makes sense to me that all Git ref types (branch, tag, commit etc.) would work the same way.

Comment (johananl) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: Huh. Looks like the problem is slashes in the ref field:

```
Error downloading modules: Error loading modules: error downloading 'https://github.com/johananl/tf-test.git?ref=my': /bin/git exited with 1: error: pathspec 'my' did not match any file(s) known to git
```

**Review by rata — COMMENTED** [review]
I think the overall diff looks good to me. Added some small comments. Thanks for tackling such a big PR! :)

One thing I'm confused is that it seems you created other PRs for fixes included here. The idea is to merge those and then adapt this?

On a side note, I still think we can improve the way commits are split, to be easy to review while still having a working tree in all commits (for easy bisect, etc.), but that seems out of scope for this PR and we don't have any rules on this yet.

File: packet/flatcar-linux/kubernetes/bootkube.tf
Comment (johananl) [reply packet/flatcar-linux/kubernetes/bootkube.tf]: This is a bug in the go-getter library which is used by Terraform to fetch modules - see https://github.com/hashicorp/go-getter/pull/228.

File: bare-metal/flatcar-linux/kubernetes/profiles.tf
Comment (mauriciovasquezbernal) [reply bare-metal/flatcar-linux/kubernetes/profiles.tf]: This part is a hack for https://github.com/hashicorp/terraform/issues/17251. That issue appears to be solved in Terraform 0.12, so probably the hack is not needed anymore.  However I avoided to touch that part and prefer to do later on.

File: packet/flatcar-linux/kubernetes/require.tf
Comment (johananl) [inline packet/flatcar-linux/kubernetes/require.tf:28]: We should update the docs as well.

File: bare-metal/flatcar-linux/kubernetes/profiles.tf
Comment (rata) [reply bare-metal/flatcar-linux/kubernetes/profiles.tf]: Oh, okay. Got confused by that only the `TODO(Mauricio)` part was added in this PR, not the rest of the comment.

btw, what does "TODO(Mauricio)" means? Didn't know that convention

**Comment by mauriciovasquezbernal** [comment]
> I think the overall diff looks good to me. Added some small comments. Thanks for tackling such a big PR! :)
> 
> One thing I'm confused is that it seems you created other PRs for fixes included here. The idea is to merge those and then adapt this?
> 

Yes, while working on this I found some problems in Azure, since those are not related to this PR I opened a separated PR for that https://github.com/kinvolk/lokomotive-kubernetes/pull/116. That's now merged and I'll rebase it to include those changes.

> On a side note, I still think we can improve the way commits are split, to be easy to review while still having a working tree in all commits (for easy bisect, etc.), but that seems out of scope for this PR and we don't have any rules on this yet.

The problem is that in order to do that I'll have to squash the commits where terraform 0.12upgrade and the fixes for that are done, if that's not a problem I can do that.

I think the following layout could work:
- preparatory commits: update provider versions, rename count to worker_count
- one commit per provider: run terraform 0.12upgrade, fixes problems, point bootkube to right place, update doc syntax
- use array notation
- use correct datatypes

Thoughts @rata, @invidian , @johananl ?

**Comment by rata** [comment]
> > I think the overall diff looks good to me. Added some small comments. Thanks for tackling such a big PR! :)
> > One thing I'm confused is that it seems you created other PRs for fixes included here. The idea is to merge those and then adapt this?
> 
> Yes, while working on this I found some problems in Azure, since those are not related to this PR I opened a separated PR for that #116. That's now merged and I'll rebase it to include those changes.

Cool! :)

> > On a side note, I still think we can improve the way commits are split, to be easy to review while still having a working tree in all commits (for easy bisect, etc.), but that seems out of scope for this PR and we don't have any rules on this yet.
> 
> The problem is that in order to do that I'll have to squash the commits where terraform 0.12upgrade and the fixes for that are done, if that's not a problem I can do that.
> 
> I think the following layout could work:
> 
> * preparatory commits: update provider versions, rename count to worker_count
> * one commit per provider: run terraform 0.12upgrade, fixes problems, point bootkube to right place, update doc syntax
> * use array notation
> * use correct datatypes
> 
> Thoughts @rata, @invidian , @johananl ?

That sound good to me, if it doesn't take too much time for you. It is, in a way, what I was trying to say here (haven't realized about the "after" changes, haven't review those probably at that time): https://github.com/kinvolk/lokomotive-kubernetes/pull/119#pullrequestreview-336819228. Sorry I wasn't clear there :-/

But I'll let others decide on this :)

File: packet/flatcar-linux/kubernetes/controllers.tf
Comment (invidian) [reply packet/flatcar-linux/kubernetes/controllers.tf:54]: I'm already afraid of `0.16upgrade` :joy:

File: aws/flatcar-linux/kubernetes/workers/variables.tf
Comment (invidian) [reply aws/flatcar-linux/kubernetes/workers/variables.tf:25]: I think that should be done in separate PR.

File: docs/flatcar-linux/kvm-libvirt.md
Comment (rata) [reply docs/flatcar-linux/kvm-libvirt.md:81]: Ohh, cool. Didn't notice!

File: aws/flatcar-linux/kubernetes/bootkube.tf
Comment (invidian) [reply aws/flatcar-linux/kubernetes/bootkube.tf]: We refer to the branch just for testing, right? BTW, you can also refer to the commit, which is not part of the master.

Before merging, that should be changed to point to the merge commit from bootkube master, so we pin it to specific version.

**Comment by invidian** [comment]
>preparatory commits: update provider versions, rename count to worker_count

It can be separated PR to minimize the impact of this PR (it's already large).

>one commit per provider: run terraform 0.12upgrade, fixes problems, point bootkube to right place, update doc syntax

IMO all providers at a time is fine. If we decide to commit per provider, we should be consistent around that with further changes. Given that it's unlikely that this commits will be reverted, I think it should be fine to have one commit for all providers.

Also the reasoning etc is the same for all providers, which is another argument for doing that together.

**Review by johananl — COMMENTED** [review]
I've tested Packet and it worked fine minus one fix to the docs (see inline comment).

I haven't reviewed the code changes yet.

File: aws/flatcar-linux/kubernetes/bootkube.tf
Comment (mauriciovasquezbernal) [reply aws/flatcar-linux/kubernetes/bootkube.tf]: I preferred to point to a branch because I can force push to https://github.com/kinvolk/terraform-render-bootkube/pull/33 without having to update the reference in this PR.

After https://github.com/kinvolk/terraform-render-bootkube/pull/33 is merged I'll update this PR to point to a single commit in the master branch.

File: packet/flatcar-linux/kubernetes/controllers.tf
Comment (mauriciovasquezbernal) [reply packet/flatcar-linux/kubernetes/controllers.tf:54]: Not as much as I am. :rofl:

File: packet/flatcar-linux/kubernetes/require.tf
Comment (mauriciovasquezbernal) [reply packet/flatcar-linux/kubernetes/require.tf:28]: Done!

File: bare-metal/flatcar-linux/kubernetes/profiles.tf
Comment (mauriciovasquezbernal) [reply bare-metal/flatcar-linux/kubernetes/profiles.tf]: Actually it was never my intention to push it, it was like a note to myself. Will remove my name but leave the TODO.

File: aws/flatcar-linux/kubernetes/variables.tf
Comment (mauriciovasquezbernal) [reply aws/flatcar-linux/kubernetes/variables.tf:113]: Terraform does the number <-> string conversion automatically: https://www.terraform.io/docs/configuration/types.html#conversion-of-primitive-types.

Comment (johananl) [reply aws/flatcar-linux/kubernetes/variables.tf:113]: Yes, I was wondering about the same thing and I've verified this works properly.

Comment (rata) [reply aws/flatcar-linux/kubernetes/variables.tf:113]: I see the PR that changes this to number type. Closing

Comment (rata) [reply aws/flatcar-linux/kubernetes/variables.tf:113]: https://github.com/kinvolk/terraform-render-bootkube/pull/33/files#diff-c9ac8098c5ea9d3e6a9a596ff0c512a4R48

**Comment by invidian** [comment]
Rebased on latest master and pushed a commit, which changes CI to use Terraform 0.12.

**Comment by invidian** [comment]
Had to cherry-pick some commits from #112 to resolve version skew between kube-apiserver and kubelets.

**Comment by pothos** [comment]
I cherry-picked one more commit from https://github.com/kinvolk/lokomotive-kubernetes/pull/112 since it seems to be related to the bootkube failure. I adapted it to apply to baremetal as well.

**Comment by invidian** [comment]
Let's merge #112 first, what do you think?

**Comment by pothos** [comment]
Yes, so that we can drop the commits here again instead of trying to keep them in sync.

**Review by invidian — APPROVED** [review]

**Comment by invidian** [comment]
Given that it took us so much time to get there, I'll merge it now.

**Pull request merged**
